src/main/java/com/exasol/adapter/document/files/FilesDocumentFetcherFactory.java

```diff
@@ -1,36 +1,104 @@
+     * @param fileLoaderFactory               dependency injection of {@link FileLoaderFactory}
+     * @param connectionInformation           connection information
+     * @param fileTypeSpecificDocumentFetcher file type specific document fetcher
      * @return built {@link DocumentFetcher}
      */
     public List<DocumentFetcher> buildDocumentFetcherForQuery(final StringFilter sourceFilter,
-            final int maxNumberOfParallelFetchers, final FileLoaderFactory fileLoaderFactory);
+            final int maxNumberOfParallelFetchers, final FileLoaderFactory fileLoaderFactory,
+            final ExaConnectionInformation connectionInformation,
+            final FileTypeSpecificDocumentFetcher fileTypeSpecificDocumentFetcher) {
+        final List<DocumentFetcher> documentFetchers = new ArrayList<>(maxNumberOfParallelFetchers);
+        final int numberOfSegments = sourceFilter.hasWildcard() ? maxNumberOfParallelFetchers : 1;
+        final StringFilter filterWithPreventedInjection = addPrefixToFilterToAvoidInjection(sourceFilter,
+                connectionInformation);
+        final List<SegmentDescription> segmentDescriptions = buildSegmentDescriptions(fileLoaderFactory,
+                connectionInformation, numberOfSegments, filterWithPreventedInjection);
+        for (final SegmentDescription segmentDescription : segmentDescriptions) {
+            final DocumentFetcher documentFetcher = new FilesDocumentFetcher(filterWithPreventedInjection,
+                    segmentDescription, fileLoaderFactory, fileTypeSpecificDocumentFetcher);
+            documentFetchers.add(documentFetcher);
+        }
+        return documentFetchers;
+    }
+
+    private StringFilter addPrefixToFilterToAvoidInjection(final StringFilter sourceFilter,
+            final ExaConnectionInformation connectionInformation) {
+        final String prefix = connectionInformation.getAddress();
+        final StringFilter filePatternWithPrefix = new PrefixPrepender().prependStaticPrefix(prefix, sourceFilter);
+        return new StringFilterFactory().and(filePatternWithPrefix, WildcardExpression.forNonWildcardPrefix(prefix));
+    }
+
+    private List<SegmentDescription> buildSegmentDescriptions(final FileLoaderFactory fileLoaderFactory,
+            final ExaConnectionInformation connectionInformation, final int numberOfSegments,
+            final StringFilter filePattern) {
+        final FileLoader loader = fileLoaderFactory.getLoader(filePattern, new NoSegmentationSegmentDescription(),
+                connectionInformation);
+        final Iterator<RemoteFile> iterator = loader.loadFiles();
+        final List<RemoteFile> firstFiles = new ArrayList<>();
+        int remoteFileCounter = 0;
+        while (iterator.hasNext() && remoteFileCounter < MAX_NUMBER_OF_FILES_TO_DISTRIBUTE_EXPLICITLY) {
+            firstFiles.add(iterator.next());
+            remoteFileCounter++;
+        }
+        if (iterator.hasNext()) {
```

Comment:
One questions about this part. 
Here the iterator can have more elements, when it is larger than `MAX_NUMBER_OF_FILES_TO_DISTRIBUTE_EXPLICITLY`, then we use hash based segmentation which will load files again. Could you please confirm?

Reply:
Exactly. In  that case each UDF will list all files and decide which to load based on hash sum.

Reply:
Thanks, got it.